src/nvim/getchar.c

```diff
@@ -1707,620 +1706,610 @@ static int vgetorpeek(const int advance)
+              while (showcmd_len < typebuf.tb_len)
+                (void)add_to_showcmd(typebuf.tb_buf[typebuf.tb_off
+                    + showcmd_len++]);
+              curwin->w_wcol = old_wcol;
+              curwin->w_wrow = old_wrow;
             }
 
-            /*
-             * get a character: 3. from the user - get it
-             */
-            wait_tb_len = typebuf.tb_len;
-            c = inchar(
-                typebuf.tb_buf + typebuf.tb_off + typebuf.tb_len,
-                typebuf.tb_buflen - typebuf.tb_off - typebuf.tb_len - 1,
-                advance ? calc_waittime(keylen) : 0,
-                typebuf.tb_change_cnt);
-
-            if (showcmd_len != 0)
-              pop_showcmd();
-            if (c1 == 1) {
-              if (State & INSERT)
-                edit_unputchar();
-              if (State & CMDLINE)
-                unputcmdline();
-              else
-                setcursor();                /* put cursor back where it belongs */
+            /* this looks nice when typing a dead character map */
+            if ((State & CMDLINE)
+                && cmdline_star == 0
+                && ptr2cells(typebuf.tb_buf + typebuf.tb_off
+                  + typebuf.tb_len - 1) == 1) {
+              putcmdline(typebuf.tb_buf[typebuf.tb_off
+                  + typebuf.tb_len - 1], FALSE);
+              c1 = 1;
             }
           }
 
-          if (c < 0)
-            continue;                     /* end of input script reached */
-          if (c == NUL) {                 /* no character available */
-            if (!advance) {
-              exiting = true;
-              break;
-            }
-            if (wait_tb_len > 0) {                /* timed out */
-              timedout = TRUE;
-              continue;
-            }
-          } else {          /* allow mapping for just typed characters */
-            while (typebuf.tb_buf[typebuf.tb_off
-                + typebuf.tb_len] != NUL)
-              typebuf.tb_noremap[typebuf.tb_off
-                + typebuf.tb_len++] = RM_YES;
+          /*
+           * get a character: 3. from the user - get it
+           */
+          wait_tb_len = typebuf.tb_len;
+          c = inchar(
+              typebuf.tb_buf + typebuf.tb_off + typebuf.tb_len,
+              typebuf.tb_buflen - typebuf.tb_off - typebuf.tb_len - 1,
+              advance ? calc_waittime(keylen) : 0,
+              typebuf.tb_change_cnt);
+
+          if (showcmd_len != 0)
+            pop_showcmd();
+          if (c1 == 1) {
+            if (State & INSERT)
+              edit_unputchar();
+            if (State & CMDLINE)
+              unputcmdline();
+            else
+              setcursor();                /* put cursor back where it belongs */
           }
         }
-      }             /* for (;;) */
-    }           /* if (!character from stuffbuf) */
 
-    /* if advance is FALSE don't loop on NULs */
-  } while (!exiting);
```

Comment:
This check used to be `c < 0 || (advance && c == NUL)` and `c` is mutated in this loop body. Guaranteeing `exiting` will be set correctly is really hard. You may have done it correctly here, but it's getting much harder for someone to change this code in the future.